sds_data_manager/constructs/sdc_step_function.py

```diff
@@ -0,0 +1,116 @@
+        instrument_task = tasks.LambdaInvoke(self,
+                                             f"InstrumentLambda-{processing_step_name}",
+                                             lambda_function=processing_system.instrument_lambda,
+                                             payload=sfn.TaskInput.from_object(
+                                                 {"TIMEOUT_TIME.$": "$.TIMEOUT_TIME"}),
+                                             result_path="$.InstrumentOutput",
+                                             result_selector={
+                                                 "STATE.$": "$.Payload.STATE",
+                                                 "JOB_NAME.$": "$.Payload.JOB_NAME",
+                                                 "COMMAND.$": "$.Payload.COMMAND",
+                                                 "OUTPUT_PATH": "$.Payload.OUTPUT_PATH",
+                                                 "INSTRUMENT_TARGET":
+                                                     "$.Payload.INSTRUMENT_TARGET"
+                                             })
+
+        # Batch Job Inputs
+        stack = Stack.of(self)
+        job_definition_arn = \
+            f'arn:aws:batch:{stack.region}:{stack.account}:job-definition/' \
+            f'{batch_resources.job_definition_name}'
+        job_queue_arn = f'arn:aws:batch:{stack.region}:{stack.account}:job-queue/' \
+                        f'{batch_resources.job_queue_name}'
+
+        instrument_target = f"{instrument_target}"
+
+        # Batch Job
+        submit_job = tasks.BatchSubmitJob(
+            self, f"BatchJob-{processing_step_name}",
+            job_name=sfn.JsonPath.string_at("$.InstrumentOutput.JOB_NAME"),
+            job_definition_arn=job_definition_arn,
+            job_queue_arn=job_queue_arn,
+            container_overrides=tasks.BatchContainerOverrides(
+                command=sfn.JsonPath.list_at("$.InstrumentOutput.COMMAND"),
+                environment={
+                    "OUTPUT_PATH": archive_bucket.bucket_name,
+                    "INSTRUMENT_TARGET": instrument_target
+                }
+            ),
+            result_path='$.BatchJobOutput'
+        )
+
+        # Success and Fail Final States
+        fail_state = sfn.Fail(self, "Fail State")
+
+        # Choice State
+        instrument_status = sfn.Choice(self, "Success?")
+        # Go to Batch job
+        created = sfn.Condition.string_equals("$.InstrumentOutput.STATE", "SUCCESS")
+        instrument_status.when(created, submit_job)
+        instrument_status.otherwise(fail_state)
+
+        submit_job.add_catch(fail_state)
+
+        # State sequences
+        add_specifics_to_input.next(
+            instrument_task).next(
+            instrument_status)
+
+        # Define the state machine
+        definition_body = sfn.DefinitionBody.from_chainable(add_specifics_to_input)
+        self.state_machine = sfn.StateMachine(self,
```

Comment:
So we get a state from the lambda, a batch job is created (which will run the container?), then that is all put into the state machine which is run in this line?

Reply:
Yes. The Batch job will run the container from the ECR (and there will be a different container for each instrument)